ansible-scylla-node/tasks/upgrade/main.yml

```diff
@@ -186,120 +196,160 @@
+- name: Get versions of {{ scylla_edition }} package
+  shell: apt list -a {{ scylla_package_prefix }} 2>/dev/null | grep `dpkg --print-architecture` | awk '{split($0,a," "); print a[2]}' | egrep -v "^{{ scylla_version_escaped }}[0123456789]+" | egrep "^{{ scylla_version_escaped }}" | sort | uniq
+  register: aptversions
+  vars:
+    scylla_version_escaped: "{{ scylla_version_to_install | regex_escape }}"
+  when: ansible_os_family == 'Debian'
 
-- name: Trigger post-upgrade tasks
-  include_tasks: post_upgrade.yml
+- name: "Validate scylla version correctness"
+  ansible.builtin.fail:
+    msg: "Too many/few choices for a requested version '{{ scylla_version_to_install }}': {{ aptversions.stdout_lines }}. Bailing out!"
+  when: ansible_os_family == 'Debian' and aptversions.stdout_lines | length != 1 and scylla_version_to_install != 'latest'
+  any_errors_fatal: True
 
+- name: Fetch version parts of Scylla package
+  set_fact:
+    scylla_version_split: "{{ aptversions.stdout | regex_findall(regexp, ignorecase=True) }}"
+  vars:
+    # All we know that the version is a string comprised of 3 parts separated by '-'
+    regexp: '^([^\-]+)-([^\-]+)-([^\-]+)$'
+  when: ansible_os_family == 'Debian'
+
+- name: Append results to list of packages
+  set_fact:
+    pkg_list: "{{ pkg_list + [ item.key + '=' + aptversions.stdout ] }}"
+  loop: "{{ installed_pkgs | dict2items }}"
+  when: not scylla_upgrade['upgrade'] and ansible_os_family == 'Debian'
+
+# Upgrade
+- name: Trigger pre-upgrade tasks
+  run_once: True
+  include_tasks: upgrade.yml
+  loop: "{{ groups['scylla'] }}"
+  loop_control:
+    loop_var: delegate_host
 
 # Upgrade verification
-- name: Upgrade verification
-  block:
-    - name: "Wait {{ upgrade_break_before_verification }} seconds before verifying the upgrade"
-      ansible.builtin.wait_for:
-        timeout: "{{ upgrade_break_before_verification |int }}"
-
-    - name: Verify if the upgrade was done sucessfully
-      include_tasks: node_verification.yml
-  rescue:
-    # Rollback
-    - name: Trigger pre-rollback tasks
-      include_tasks: pre_rollback.yml
-
-    #- name: Trigger operating system specific rollback tasks
-    #  include_tasks: "rollback_{{ scylla_support[ansible_facts['distribution'] |lower][ansible_facts['distribution_major_version']]['template'] }}.yml"
-
-    #- name: Trigger post-rollback tasks
-    #  include_tasks: post_rollback.yml
-    
-    # Rollback verification
-    - name: Wait {{ upgrade_break_before_verification }} seconds before verifying the rollback
-      ansible.builtin.wait_for:
-        timeout: "{{ upgrade_break_before_verification |int }}"
-
-    - name: Verify if the upgrade was done sucessfully
-      include_tasks: node_verification.yml
-  
-
-  # ----- TODO: If the upgrade version is the latest one and if 'latest' was used to specify the version
-  # replace upgrade version with latest
+- name: "Wait {{ upgrade_break_before_verification }} seconds before verifying the upgrade"
```

Comment:
The whole section before this one is quite messy IMO. 

Please, correct me if I'm wrong but the section above does following logical steps:

1) Gets the full version and its parts and checks if upgrade is possible/valid:
  a) If the requested version is not greater than the "latest" - not needed if you do (b) below.
  b) (It should but you don't do that yet AFAIKT, at least for RH you don't): check that the requested version exists at all (my code for Debian you picked does that).
  c) Check if we do a switch to a different major version and only allow this if a special variable is set (to avoid accidental major upgrades).
2) Check if any action is needed: if the upgrade version is equal to the already installed version. In your patch you seem to only check this for 'latest' however the same logic applies to any version. 

So, the way the code is now (before and with your patch) it's very flat and all mixed up: Debian and RH sections are interleaving and appear in a flat (not structured) way. This not only makes it hard to read but also makes it hard to check because same operations and checks need to be done for both OS variants.

It would make sense IMO to rewrite the code the way that the logic described above would be implemented in the abstract way with minimum per-variant "branching". 

The generic logic (like comparing that the requested and the installed versions are the same) should be written in a generic way and rely on `facts` initialized by a variant-specific code.

variant-specific code blocks should be taken into separate files and then "included" from the main.yml.

This way it would be much easier to both review and maintain such a code.

Reply:
Please, review the whole patch with the mindset described above.

It's probably a lot of work but I believe it's worth it - it's going to make life of all that would come after us MUCH easier.

Reply:
>  b) (It should but you don't do that yet AFAIKT, at least for RH you don't): check that the requested version exists at all (my code for Debian you picked does that).

We do check for it. `latest_detected_release` will return `latest`. Then, we can infer that any upgrade should be between our current version and whatever latest is. That check is done under:
```yaml
   when: not upgrade_major and scylla_upgrade['version'] is version(latest_detected_release,'>=') and scylla_version != 'latest'
```

That means that:
- If the user specified **latest**, we will fail if he's already latest in the subsequent check.
- If the user specified **a version**, then:
  - If the version is >= latest, we fail
  - If the version == current, we fail 
  - If the version < current, we force him to specify that he wants to downgrade

Other than that, your understanding at 1. is correct.

> 2. Check if any action is needed: if the upgrade version is equal to the already installed version. In your patch you seem to only check this for 'latest' however the same logic applies to any version.

This check actually happens in line 204. Which is when the user has specified a version other than `latest`. :-)

Other than that, yes - I agree that its all messy and scattered around. Splitting it is actually fairly simple.